src/remotes/auth0/index.js

```diff
@@ -62,45 +65,109 @@ function getConnectionResolvers(prefix, schemas) {
   };
   resolvers.Mutation = {
     updateUser: async (_, { username, updates }, ctx) => {
-      await updateUser(username, ctx, (prev) => ({
-        ...prev,
-        ...updates
-      }));
+      await updateUser(username, ctx, (prev) => {
+        const newUser = {
+          ...prev,
+          ...updates
+        }
+        newUser.name = (updates.displayNameFormat ? formatName(newUser.displayNameFormat, newUser.givenName, newUser.familyName) : newUser.name)
+        return newUser;
+      });
     },
-    grantBadge: async (_, { username, badge }, ctx) => {
-      await updateUser(username, ctx, (prev) => ({
+    updateAuthUser: async (_, { updates }, ctx) => {
+      if (!ctx.user) {
+        throw Error("Sorry! You don't have permission to do this. Please refresh the page and try again.")
+      } else if (updates.username && updates.username !== updates.username.replace(/[^a-zA-Z0-9\-_]/g, '')) {
+        throw new Error('Username can only consist of letters, numbers, and _ or -.');
+      } else if (!updates.familyName && typeof updates.familyName !== 'undefined') {
+        throw new Error('Name is required.');
+      } else if (!updates.givenName && typeof updates.givenName !== 'undefined') {
+        throw new Error('Name is required.');
+      }
+      await updateUser(null, ctx, (prev) => {
+        const newUser = {
+          ...prev,
+          ...updates
+        }
+        newUser.name = formatName(newUser.displayNameFormat, newUser.givenName, newUser.familyName)
+        return newUser;
+      });
+    },
+    grantBadge: async (_, { where, badge }, ctx) => {
+      await updateUser(where, ctx, (prev) => ({
         ...prev,
         badges: [
           ...(prev.badges || []).filter((b) => b.id !== badge.id),
           badge,
         ]
       }));
+    },
+    revokeBadge: async (_, { where, badge }, ctx) => {
+      await updateUser(where, ctx, (prev) => ({
+        ...prev,
+        badges: [
+          ...(prev.badges || []).filter((b) => b.id !== badge.id)
+        ]
+      }));
+    },
+    uploadPicture: async (_, { upload }, ctx) => {
+      const { createReadStream, filename } = await upload;
+      console.log(filename)
+      const chunks = [];
+      // eslint-disable-next-line no-restricted-syntax
+      for await (const chunk of createReadStream()) {
+        chunks.push(chunk);
+      }
+      const uploadBuffer = Buffer.concat(chunks);
+
+      const result = await uploader.image(uploadBuffer, filename || '_.jpg');
+
+      return result.url
+    },
+    addRole: async (_, {id, roleId}, ctx) => {
+      addRole(id, roleId, ctx)
     }
   }
   const lru = new LruCache({ maxAge: 1000 * 60 * 5, max: 500 });
   resolvers.User = {
-    roles: async ({ id }, _, ctx) => requireScope(ctx, scopes.readUserRoles) && getRolesForUser(id),
+    roles: async ({ id }, _, ctx) => {
+      if (!hasScope(ctx, scopes.readUserRoles) && !hasScope(ctx, `read:${ctx.user}`)) {
```

Comment:
This new scope should be named `read:users:${user}` instead of `read:${user}`, because `read:${user}` would allow someone to inject any subkey of `read` by choosing the right username.

(e.g. if I picked the username "users", this scope would be set to `read:users` which would let me read any user)